File: medium/characterReplacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        char_freq = dict()
        max_freq = 0
        start = 0
        max_length = 0
        n = len(s)
        for end in range(n):
            char_freq[s[end]] = char_freq.get(s[end], 0) + 1
            max_freq = max(max_freq, char_freq[s[end]])
            while (end - start + 1) - max_freq > k:
                char_freq[s[start]] -= 1
                start += 1
            max_length = max(max_length, end - start + 1)
        return max_length


def main():
    sol = Solution()
    s = "AABABBA"
    k = 1
    ans = sol.characterReplacement(s, k)
    print(ans)


if __name__ == "__main__":
    main()

File: medium/characterReplacement.js
class Solution {
  characterReplacement(s, k) {
    let count = {};
    let ans = 0;
    let l = 0;

    for (let r = 0; r < s.length; r++) {
      count[s[r]] = 1 + (count[s[r]] || 0);
      if (r - l + 1 - Math.max(...Object.values(count)) > k) {
        count[s[l]]--;
        l++;
      }
      ans = Math.max(ans, r - l + 1);
    }

    return ans;
  }
}
